apis/appmesh/v1beta2/virtualservice_types.go

```diff
@@ -17,25 +17,79 @@ limitations under the License.
 package v1beta2
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
-// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+// VirtualNodeServiceProvider refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_VirtualNodeServiceProvider.html
+type VirtualNodeServiceProvider struct {
+	// The virtual node that is acting as a service provider.
+	VirtualNodeRef VirtualNodeReference `json:"virtualNodeRef"`
+}
+
+// VirtualRouterServiceProvider refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_VirtualRouterServiceProvider.html
+type VirtualRouterServiceProvider struct {
+	// The virtual router that is acting as a service provider.
+	VirtualRouterRef VirtualRouterReference `json:"virtualRouterRef"`
+}
+
+// VirtualServiceProvider refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_VirtualServiceProvider.html
+type VirtualServiceProvider struct {
+	// The virtual node associated with a virtual service.
+	// +optional
+	VirtualNode *VirtualNodeServiceProvider `json:"virtualNode,omitempty"`
+	// The virtual router associated with a virtual service.
+	// +optional
+	VirtualRouter *VirtualRouterServiceProvider `json:"virtualRouter,omitempty"`
+}
+
+type VirtualServiceConditionType string
+
+const (
+	// VirtualServiceActive is True when all the AppMesh VirtualService has been created or found via the API
+	VirtualServiceActive VirtualServiceConditionType = "VirtualServiceActive"
+)
+
+type VirtualServiceCondition struct {
+	// Type of VirtualService condition.
+	Type VirtualServiceConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status"`
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`
+	// The reason for the condition's last transition.
+	// +optional
+	Reason *string `json:"reason,omitempty"`
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message *string `json:"message,omitempty"`
+}
 
 // VirtualServiceSpec defines the desired state of VirtualService
 type VirtualServiceSpec struct {
```

Comment:
Don't we need an 'awsName' field for VirtualService as well since we create an AWS App Mesh object for this as well?

Reply:
it's will be the hosts. an k8s virtualService corresponding to multiple appmesh virtualService.
each host will become a appMesh virtualService's awsName :D

Reply:
Ok, so are we going to create multiple Virtual Service App Mesh objects based on hosts array for a single K8S Virtual Service? Let's say 1 for "svc-1.my-ns.svc.cluster.local" and 1 for "svc-1" and so on...?

Reply:
yes :D.
and by default, the host array will only have one entry, which is the VirtualService CRD's name.

Reply:
FYI, i changed it to just be 'awsName' instead of 'hosts' array, per https://github.com/aws/aws-app-mesh-controller-for-k8s/pull/150#discussion_r404365334